src/coreclr/src/interop/inc/interoplib.h

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _INTEROP_INC_INTEROPLIB_H_
+#define _INTEROP_INC_INTEROPLIB_H_
+
+namespace InteropLibImports
+{
+    // Forward declaration of Runtime calling context.
+    // This class is used by the consuming runtime to pass through details
+    // that may be required during a subsequent callback from the InteropLib.
+    // InteropLib never directly modifies or inspects supplied instances.
+    struct RuntimeCallContext;
+}
+
+namespace InteropLib
+{
+    using OBJECTHANDLE = void*;
+
+    // Clean up the interop library.
+    void Shutdown() noexcept;
+
+#ifdef _WIN32
+
+    namespace Com
+    {
+        // Create an IUnknown instance that represents the supplied managed object instance.
+        HRESULT CreateWrapperForObject(
+            _In_ OBJECTHANDLE instance,
+            _In_ INT32 vtableCount,
+            _In_ void* vtables,
+            _In_ INT32 flags,
+            _Outptr_ IUnknown** wrapper) noexcept;
+
+        // Destroy the supplied wrapper
+        void DestroyWrapperForObject(_In_ void* wrapper) noexcept;
+
+        struct ExternalWrapperResult
+        {
+            // The returned context memory is guaranteed to be initialized to zero.
+            void* Context;
+
+            // IAgileReference instance.
+            IUnknown* AgileRef;
+
+            // See https://docs.microsoft.com/windows/win32/api/windows.ui.xaml.hosting.referencetracker/ for
+            // details.
+            bool FromTrackerRuntime;
+        };
+
+        // Allocate a wrapper context for an external object.
+        // The runtime supplies the external object, flags, and a memory
+        // request in order to bring the object into the runtime.
+        HRESULT CreateWrapperForExternal(
+            _In_ IUnknown* external,
+            _In_ INT32 flags,
+            _In_ size_t contextSize,
+            _Out_ ExternalWrapperResult* result) noexcept;
+
+        // Destroy the supplied wrapper.
+        void DestroyWrapperForExternal(_In_ void* context) noexcept;
+
+        // Separate the supplied wrapper from the tracker runtime.
+        void SeparateWrapperFromTrackerRuntime(_In_ void* context) noexcept;
+
+        // Register a runtime implementation callback in the Reference Tracker Host scenario.
+        // Returns true if registration succeeded, otherwise false.
+        bool RegisterReferenceTrackerHostRuntimeImpl(_In_ OBJECTHANDLE objectHandle) noexcept;
```

Comment:
This is a singleton anyway. The managed reference can be just a field in C#. I do not think it needs to be wired all the way into the interop lib just so that it can be passed all the way back.

Reply:
I typically try to avoid using `NULL` as anything other than a predicate invariant which is why I passed the object around. However, this seems like a reasonable idea to avoid creating yet _another_ object handle.

Reply:
Continuing the thought process here. Letting `NULL` through does makes some of these code paths less rigorous since now `NULL` is permitted and the state is more magical. Is there a strong desire to remove this registration? I am not against it but it does loosen so many of the contracts and increases the mental/testing complexity of understanding the system.

Reply:
I meant to delete the OBJECTHANDLE argument. I did not meant to pass NULL.

Reply:
Agreed. I can remove the registration, but the usage of this value originates from within the InteropLib via a call from Jupiter. Which means I would need to pass null from there all the way back up to indicate use the static field on the `ComWrappers` rather than being provided one.

Reply:
Hmmmm perhaps there is another interpretation for your statement. Is there an invalid `OBJECTREF` value that I can use as an invariant to use?

Reply:
Done.